src/main/java/sparksoniq/spark/DataFrameUtils.java

```diff
@@ -146,4 +159,31 @@ public static Object deserializeRowColumn(Row row, int columnIndex) {
         KM.releaseKryoInstance(kryo);
         return deserializedColumnObjects;
     }
+
+    public static Dataset<Row> zipWithIndex(Dataset<Row> df, Long offset, String indexName) {
+        Dataset<Row> dfWithPartitionId = df
+                .withColumn("partition_id", spark_partition_id())
+                .withColumn("inc_id", monotonically_increasing_id());
+
+        Object partitionOffsetsObject = dfWithPartitionId
+                .groupBy("partition_id")
+                .agg(count(lit(1)).alias("cnt"), first("inc_id").alias("inc_id"))
+                .orderBy("partition_id")
+                .select(col("partition_id"), sum("cnt").over(Window.orderBy("partition_id")).minus(col("cnt")).minus(col("inc_id")).plus(lit(offset).alias("cnt")))
```

Comment:
It would be great if this could be indented and arranged over multiple lines to make it more readable.